### PR TITLE
hash bench

### PR DESCRIPTION
**Pull request by hackaugusto**
## Describe your changes

A few changes to the hash benchmarks done while reviewing #76

Ref: https://github.com/0xPolygonMiden/crypto/pull/76#issuecomment-1444152014

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: benches/hash.rs [hunk 1]
```diff
@@ -1,116 +1,110 @@
...
-        bench.iter_batched(
-            || {
-                [
-                    Blake3_256::hash(&rand_value::<u64>().to_le_bytes()),
-                    Blake3_256::hash(&rand_value::<u64>().to_le_bytes()),
-                ]
+    // Benchmark `merge` in a hot loop with varying number of elements, the goal is to spend more
+    // time in the hashing function so that the cost of measuring it will dissipiate in the
+    // background and give better view of the algorithms performance.
+    for size in BATCH_SIZES {
```
Comment (bobbinth) [inline benches/hash.rs:50]: I think it is fine if we test it on just a couple of sizes - e.g, 64 and 256.

File: benches/hash.rs [hunk 2]
```diff
@@ -1,116 +1,110 @@
...
-        )
-    });
+        );
+
+        let v2: Vec<[<Blake3_256 as Hasher>::Digest; 2]> = v
+            .iter()
+            .map(|[l, r]| [l.as_bytes().into(), r.as_bytes().into()])
+            .collect();
+        group.bench_with_input(
+            BenchmarkId::new(format!("Blake3/256/{size}"), size),
```
Comment (bobbinth) [inline benches/hash.rs:76]: I think doing `/{size}` is redundant here. We could probably just do: `BenchmarkId::new("Blake3_256", size)`.

File: benches/hash.rs [hunk 3]
```diff
@@ -1,116 +1,110 @@
...
+        let v: Vec<[RpoDigest; 2]> = (0..size)
+            .map(|_| {
+                let left = rand_array::<Felt, 4>().into();
+                let right = rand_array::<Felt, 4>().into();
+                [left, right]
+            })
+            .collect();
+
+        group.bench_with_input(
+            BenchmarkId::new(format!("RPO256/{size}"), size),
```
Comment (bobbinth) [inline benches/hash.rs:60]: I think doing `/{size}` is redundant here. We could probably just do: `BenchmarkId::new("RPO256", size)`.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! I left some small comment inline.

I guess the bigger question is whether these show results which are any different from what we have [here](https://github.com/0xPolygonMiden/crypto/tree/main/benches#comparison-and-instructions). If so, we'll need to figure out what to understand what drives the discrepancy and potentially re-run the benchmarks on various platforms to update the table.

File: benches/hash.rs [hunk 4]
```diff
@@ -1,116 +1,110 @@
...
+static BATCH_SIZES: [usize; 8] = [
+    2usize.pow(0),
+    2usize.pow(1),
+    2usize.pow(2),
+    2usize.pow(3),
+    2usize.pow(4),
+    2usize.pow(5),
+    2usize.pow(6),
+    2usize.pow(7),
+];
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply benches/hash.rs:21]: changed

File: benches/hash.rs [hunk 5]
```diff
@@ -1,116 +1,110 @@
...
-    });
-}
-
-fn blake3_2to1(c: &mut Criterion) {
-    let v: [<Blake3_256 as Hasher>::Digest; 2] =
+    let s2: [<Blake3_256 as Hasher>::Digest; 2] =
         [Blake3_256::hash(&[1_u8]), Blake3_256::hash(&[2_u8])];
-    c.bench_function("Blake3 2-to-1 hashing (cached)", |bench| {
-        bench.iter(|| Blake3_256::merge(black_box(&v)))
+    group.bench_with_input(BenchmarkId::new("Blake3/256", 1), &s2, |b, i| {
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply benches/hash.rs:41]: changed all `/` to `_`

File: benches/hash.rs (hunk 3 above)
Comment (hackaugusto) [reply benches/hash.rs:60]: removed the size of all ids

File: benches/hash.rs (hunk 1 above)
Comment (hackaugusto) [reply benches/hash.rs:50]: This request conflicts with the one above: https://github.com/0xPolygonMiden/crypto/pull/77/files#r1156632675

I kept the list above. Let me know if it should change.

Note that exercising different sizes to fill the L1/L2/L3 caches is important for reliability of the results.

File: benches/hash.rs [hunk 6]
```diff
@@ -1,116 +1,110 @@
...
-                    .try_into()
-                    .expect("should not fail");
-                v
-            },
-            |state| Blake3_256::hash_elements(&state),
-            BatchSize::SmallInput,
-        )
-    });
+    for size in BATCH_SIZES {
+        let v: Vec<Felt> = (0..size).map(|_| rand_value::<Felt>()).collect();
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply benches/hash.rs:93]: changed to `rand_vector`

**Pull request closed without merging**